### ISSUE TITLE
[HOLD for payment 2023-01-19] [Performance] Use react-native-quick-sqlite instead of asyncStorage

### ISSUE THREAD

**Issue by Szymon20000** [Expensify/App#10793]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?
reducing time spend on serialisation/deserialisation and generally bridge overhead. 

## What is the impact of this on end-users?
slower app

## List any benchmarks that show the severity of the issue
https://github.com/mrousavy/react-native-mmkv#benchmark

## Proposed solution (if any)
use react-native-mmkv for Onyx

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)
Note: These should be the same as the benchmarks collected before any changes.

## Platform:
Where is this issue occurring?

- [ ] Web
- [x] iOS
- [x] Android
- [ ] Desktop App
- [ ] Mobile Web

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by roryabraham** [comment]
Switching this to weekly

**Comment by trjExpensify** [comment]
👋 Should we update the title of this issue to be "use `react-quick-sqlite` instead of asyncStorage" and does [this PR](https://github.com/Expensify/App/pull/11066) need to be linked to this issue?

**Comment by roryabraham** [comment]
Looks like there was an email sent out recently to the SQLite team to investigate and address the performance difference between AsyncStorage and quick-sqlite

**Comment by JmillsExpensify** [comment]
Latest convo and progress is happening [in Slack](https://expensify.slack.com/archives/C035J5C9FAP/p1667503389141599).

**Comment by JmillsExpensify** [comment]
Latest updates [in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1668487356465739), though also check [the PR](https://github.com/Expensify/react-native-onyx/pull/211) where all the action is right now.

**Comment by tjferriss** [comment]
It looks like Jason's comment from last week still holds true. I see activity in [the PR](https://github.com/Expensify/react-native-onyx/pull/211) so I assume this issue is moving toward a resolution.

**Comment by roryabraham** [comment]
I'm unassigning myself from this, but there are a lot of reviews on https://github.com/Expensify/react-native-onyx/pull/211. Feels like we are pretty close there. Let's get it over the finish line!

**Comment by JmillsExpensify** [comment]
@marcaaron Mind jumping in this PR when you get a chance today. Looks like you're requested at the moment.

**Comment by JmillsExpensify** [comment]
Lots of active discussion in the linked PR.

**Comment by JmillsExpensify** [comment]
Commenting again for Melvin.

**Comment by JmillsExpensify** [comment]
@Szymon20000 following on all the activity last week in the PR, is your plan to circle back this week and address? Looks like we are hopefully close and can close this out ahead of the end of the year.

**Comment by trjExpensify** [comment]
PR is in review, adding the `Reviewing` label.

**Comment by JmillsExpensify** [comment]
Nice catch. CCing a couple of people in the linked PR. @iwiznia @marcaaron @roryabraham What's needed to get this PR to the finish line ahead of the holidays?

**Comment by JmillsExpensify** [comment]
> Get the reviews done

Alright, @yuwenmemon @kidroca @parasharrajat @thienlnam @marcaaron @mountiny @roryabraham Can you all please prioritize a review so that we can unblock this specific PR? We still have several steps left, let's dig in and get this one crossed off the list.

**Comment by kidroca** [comment]
@iwiznia 
> * Add some testing to ensure this does not break everything

Do you mean a testing strategy/steps for QA?

Since Onyx logic remains the same - it is still covered by existing tests 
We can add storage provider tests - they can ensure the correct SQL queries are sent to `quick-squlite`, but nothing more than that (e.g. not critical to do now, perhaps in a follow-up PR)

> * I assume we will need some kind of migration to move all data from one place to the other

IMO this would be very little value compared to the time we'll need to spend writing the migration

It's not a simple migration like an Onyx key getting renamed to something else

To write a migration for the storage driver it:
- the migration file (written in App) would have to (try) read values using `AsyncStorage` - meaning `AsyncStorage` have to stay as package.json dependency (forever, because the migration file stays in the source forever)
- then write/migrate data to Onyx using `Onyx.update` or raw commands 
- verify the migration would not corrupt something (which is a hard task to do)

The alternative is to write no migration, which would simply render the user logged out, and once they log in they'll have their data synchronized automatically (as @marcaaron pointed out earlier: https://github.com/Expensify/react-native-onyx/pull/211#pullrequestreview-1207209722) 

A middle ground would be to try and migrate only auth data (token, session) so the user does not need to re-login. Then anything missing should be automatically synced from the server

**Comment by iwiznia** [comment]
> Do you mean a testing strategy/steps for QA?

I mean using the new react-native-onyx version in the app and playing around with it, checking nothing breaks.

> The alternative is to write no migration, which would simply render the user logged out, and once they log in they'll have their data synchronized automatically (as @marcaaron pointed out earlier: https://github.com/Expensify/react-native-onyx/pull/211#pullrequestreview-1207209722)

Good points. I think I am ok with this.

> A middle ground would be to try and migrate only auth data (token, session) so the user does not need to re-login. Then anything missing should be automatically synced from the server

If this is easy, then let's go for it.

Ah, also, I think we should add something to delete all the existing data from the old place, no? Or would the browser clear it out eventually?

**Comment by marcaaron** [comment]
Web Browser - no as we are not messing with web storage in this change.
Native - I guess we could clear this. I'm not sure if it's worth it. Fresh install of the app would eventually clear out this data. We can always do a simple migration that just does `AsyncStorage.clear()` (but requires that we include the library forever just for this purpose).

**Comment by kidroca** [comment]
> Ah, also, I think we should add something to delete all the existing data from the old place, no? Or would the browser clear it out eventually?

Good point, I don't know - need to research this 

In any case to make migrations easier to manage problems like these we need to update the migration system a bit so that 
- migrations can be loaded on demand (e.g. `require` calls) 
- irrelevant migrations can be dropped, or at least not `required` 

This would allow use to use dependencies in migrations (e.g. AsyncStorage), and perhaps carry them for a month or 2 (using some heuristics - e.g. are there any users using a version predating the migration) after they are no longer needed we can drop them altogether

**Comment by JmillsExpensify** [comment]
@tgolen Isn't this issue done at this point?

**Comment by roryabraham** [comment]
The PR to upgrade E/App to use the new quick-sqlite based version of Onyx was deployed to staging in https://github.com/Expensify/App/issues/14188, not yet deployed to production

**Comment by mountiny** [comment]
I think there is nothing else to do here since payments are handled by different flow in this case. Should we close this issue or anything else left? 🎉

**Comment by JmillsExpensify** [comment]
@Szymon20000 Can you make sure that we handle your individual $1,000 bonus for this issue on Margelo invoice?

**Comment by JmillsExpensify** [comment]
@mountiny I'll go ahead and close the issue, though can you double confirm that the BZ checklist above doesn't apply?

**Comment by mountiny** [comment]
It does not apply in this case!
